rfcs/003-strings.md

```diff
@@ -0,0 +1,325 @@
+
+* Introduce a new type `CWString` which is the wide string equivalent to `CString`.
+* win32 strings and `BSTR`s should always be treated like raw pointers. Accessing them is unsafe since the compiler can not verify that they are still live with the exception of their use as "IN" params which we discuss below.
+* `HSTRING`s are always live and should have their reference count decremented in their `Drop` impl.
+
+### Converting between String types
+
+#### From Windows to Rust types
+
+The main distinction between `HSTRING` and the other Windows string types is that `HSTRING` is always valid while the other string types are essentially wrappers around raw pointers and thus cannot be assumed to point to valid memory. Therefore, while `HSTRING` can host a variety of safe conversions into Rust strings, the other strings cannot.
+
+The main conversions afforded to these types are to:
+
+* `&[u16]` or `&[u8]` (depending if the given Windows type is a wide string or not)
+* `String`
+
+Conversion of most types to `CString` and `OString` doesn't make much sense as all Windows types are already FFI safe and those Rust types are almost never needed outside of FFI. However, the user should be given access to the underlying buffers (including the trailing nul byte), and there should be conveniences to converting to `String` since this unlocks the entire rich `std` library functionality for string manipulation.
+
+All string types that aren't `HSTRING` can only expose unsafe functions as it is not well known that they point to valid memory. Additionally, each type should allow conversion to a raw pointer to avoid having to use `std::mem::transmute` in cases where an API expects a raw pointer.
+
+#### From Rust to Windows types
+
+It is common for Rust functionality to want to interact with some Windows APIs which require the Windows string types.
+
+Converting from Rust types usually requires copy the UTF-8 encoded bytes to UTF-16. This can be done with any type that can be referenced as a `str`, so it makes sense to provide `From<&T> where T: AsRef<str>` for `HSTRING` as `HSTRING` will ensure the newly allocated buffer is freed when appropriate.
+
+All the other Windows string types do not have clear ownership semantics and thus providing convenient conversions might risk leading the user to leak memory. This is the same reason logically borrowed types like `CStr` don't allow converting from owned types like `String`. As such, only direct conversions will be provided: conversion from `*const u8` and `*const u16`.
+
+#### API Proposal
+
+With the above in mind, here is what the API should look like:
+
+```rust
+impl HSTRING {
+    // String data without the trailing 0
+    fn as_wide(&self) -> &[u16] {}
+    // String data with the trailing 0
+    fn as_wide_with_nul(&self) -> &[u16] {}
+    fn to_string(&self) -> Result<String, FromUtf16Error> {}
```

Comment:
Note: This function collides with [`ToString::to_string`](https://doc.rust-lang.org/nightly/std/string/trait.ToString.html#tymethod.to_string), since `HSTRING` implements `Display` below.

Perhaps this could be renamed `try_to_string`?

Reply:
Good catch. That's a shame. We could use `into_string` which mirrors [`CString::into_string`](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_string) directly. I originally didn't do this as it's not exactly trivial to reuse the `String`'s buffer for the `HSTRING` buffer and thus I didn't want to take `self`.